File: julia/solvers/linear_ac_pf.jl
using LinearAlgebra
using Printf

"""
    solve_linear_ac_pf(buses, lines, generators, loads; baseMVA, v_nom)

Linearized AC Power Flow (LACPF).

Unlike DC PF (which ignores resistance and reactive power), LACPF linearises
the full AC power flow equations around the flat start (|V|=1 p.u., θ=0 rad)
using a first-order Taylor expansion. It solves simultaneously for:
  - voltage angle deviations  Δθ  [rad]
  - voltage magnitude deviations  Δ|V| [p.u.]

The linearised system (excluding slack bus) is:
    [ B'  -G'] [Δθ ]   [P_inj / baseMVA]
    [-G'  -B'] [Δ|V|] = [Q_inj / baseMVA]

where G' + jB' is the reduced nodal admittance matrix.

# Arguments
- `buses`      : vector of bus names
- `lines`      : vector of (from, to, r_pu, x_pu) — r,x in per-unit
- `generators` : Dict(bus => (P_MW, Q_MVAr))  (Q=0 if not reactive-controlled)
- `loads`      : Dict(bus => (P_MW, Q_MVAr))
- `baseMVA`    : system base (default 100 MVA)
- `v_nom`      : nominal voltage kV (informational only; r,x already p.u.)
"""
function solve_linear_ac_pf(buses, lines, generators, loads;
                              baseMVA = 100.0,
                              v_nom   = 380.0,
                              verbose = true)

    n = length(buses)

    # ── Build admittance matrix Y = G + jB ──────────────────────
    # r, x are in physical Ω → convert to p.u. via z_base = v_nom²/baseMVA
    z_base = v_nom^2 / baseMVA          # e.g. 380²/100 = 1444 Ω

    G = zeros(n, n)
    B = zeros(n, n)

    for (from, to, r, x) in lines
        r_pu  = r / z_base               # per-unit resistance
        x_pu  = x / z_base               # per-unit reactance
        denom = r_pu^2 + x_pu^2         # |z_pu|²

        g_ij  =  r_pu / denom            # conductance [p.u.]
        b_ij  =  x_pu / denom            # |susceptance| [p.u.] (positive)

        # Diagonal (shunt): sum of incident branch admittances
        G[from, from] += g_ij;  G[to, to]   += g_ij
        B[from, from] += b_ij;  B[to, to]   += b_ij   # B_ii > 0

        # Off-diagonal: negative admittance
        G[from, to]   -= g_ij;  G[to, from] -= g_ij
        B[from, to]   -= b_ij;  B[to, from] -= b_ij   # B_ij < 0
    end

    # Scale to MW units: B [MW/rad], G [MW/p.u.-V]
    G_mw = G .* baseMVA
    B_mw = B .* baseMVA

    # ── Net power injections [p.u.] ──────────────────────────────
    P = zeros(n)   # active   [MW] → will be divided by baseMVA
    Q = zeros(n)   # reactive [MVAr]

    for (bus, (p, q)) in generators
        P[bus] += p
        Q[bus] += q
    end
    for (bus, (p, q)) in loads
        P[bus] -= p
        Q[bus] -= q
    end

    # ── Remove slack bus (bus 1) → reduce to (n-1) × (n-1) ─────
    idx = 2:n    # non-slack indices

    B_r = B_mw[idx, idx]    # reduced B [MW/rad]
    G_r = G_mw[idx, idx]    # reduced G [MW/p.u.-V]
    P_r = P[idx]             # [MW]
    Q_r = Q[idx]             # [MVAr]

    # ── Assemble 2(n-1) × 2(n-1) system ─────────────────────────
    #   [ B_r  -G_r] [Δθ ]   [P_r]
    #   [-G_r  -B_r] [Δ|V|] = [Q_r]
    nm = n - 1
    A = [ B_r  -G_r ;
         -G_r  -B_r ]

    rhs = [P_r; Q_r]

    # ── Solve ────────────────────────────────────────────────────
    x_sol = A \ rhs

    Δθ_red = x_sol[1:nm]        # [rad]
    Δv_red = x_sol[nm+1:end]    # [p.u.]

    # ── Reconstruct full vectors ─────────────────────────────────
    θ   = zeros(n);      θ[idx]   = Δθ_red
    Δv  = zeros(n);      Δv[idx]  = Δv_red
    V_mag = 1.0 .+ Δv           # |V| = 1 + Δ|V|  [p.u.]

    # ── Line flows ───────────────────────────────────────────────
    P_flow = Float64[]
    Q_flow = Float64[]

    for (from, to, r, x) in lines
        r_pu  = r / z_base
        x_pu  = x / z_base
        denom = r_pu^2 + x_pu^2
        g_ij  = r_pu / denom     # conductance p.u.
        b_ij  = x_pu / denom     # |susceptance| p.u.

        Δθ_ft = θ[from] - θ[to]
        ΔV_ft = Δv[from] - Δv[to]

        # Linearised branch flows (1st order, flat start):
        #   P_ft ≈ (b_ij * Δθ_ft + g_ij * ΔV_ft) * baseMVA
        #   Q_ft ≈ (b_ij * ΔV_ft - g_ij * Δθ_ft) * baseMVA
        p_ft = ( b_ij * Δθ_ft + g_ij * ΔV_ft) * baseMVA
        q_ft = ( b_ij * ΔV_ft - g_ij * Δθ_ft) * baseMVA

        push!(P_flow, p_ft)
        push!(Q_flow, q_ft)
    end

    # ── Print results ────────────────────────────────────────────
    if verbose
        println("="^60)
        println("LINEARIZED AC POWER FLOW  (Julia)")
        println("="^60)

        println("\n1. VOLTAGE MAGNITUDES AND ANGLES:")
        @printf("%-10s %12s %12s %12s\n",
                "Bus", "|V| (p.u.)", "θ (rad)", "θ (deg)")
        println("-"^50)
        for i in 1:n
            @printf("%-10s %12.6f %12.6f %12.4f\n",
                    buses[i], V_mag[i], θ[i], θ[i]*180/π)
        end

        println("\n2. LINE FLOWS:")
        @printf("%-15s %12s %12s\n", "Line", "P (MW)", "Q (MVAr)")
        println("-"^42)
        for (k, (from, to, r, x)) in enumerate(lines)
            @printf("Bus%d → Bus%d    %12.4f %12.4f\n",
                    from, to, P_flow[k], Q_flow[k])
        end
    end

    return (
        V_mag   = V_mag,
        V_ang   = θ,
        P_flow  = P_flow,
        Q_flow  = Q_flow
    )
end

# ================================================================
#  Main: 3-bus validation against DC PF and full AC PF reference
# ================================================================
if abspath(PROGRAM_FILE) == @__FILE__

println("\n" * "="^60)
println("3-BUS VALIDATION: DC PF vs LINEARIZED AC PF vs Full AC PF")
println("="^60)

buses = ["Bus 0", "Bus 1", "Bus 2"]
lines = [(1, 2, 0.01, 0.1),
         (1, 3, 0.01, 0.1),
         (2, 3, 0.01, 0.1)]

generators = Dict(1 => (500.0, 0.0))
loads      = Dict(2 => (300.0, 0.0),
                  3 => (200.0, 0.0))

res = solve_linear_ac_pf(buses, lines, generators, loads)

println("\n" * "="^60)
println("COMPARISON: Linearized AC PF vs Reference (PyPSA full AC PF)")
println("="^60)

# PyPSA full AC PF reference values (from validation in Chapter 3)
ref_vm  = [1.000000, 0.999982, 0.999984]
ref_va  = [0.000000, -0.000185, -0.000162]
ref_pf  = [266.67, 233.34, -33.33]
ref_qf  = [0.048, 0.040, -0.002]

println("\nVoltage magnitudes [p.u.]:")
@printf("%-8s %12s %12s %12s\n", "Bus", "LACPF", "Full AC", "|Δ|")
for i in 1:3
    @printf("%-8s %12.6f %12.6f %12.2e\n",
            buses[i], res.V_mag[i], ref_vm[i],
            abs(res.V_mag[i] - ref_vm[i]))
end

println("\nVoltage angles [rad]:")
@printf("%-8s %12s %12s %12s\n", "Bus", "LACPF", "Full AC", "|Δ|")
for i in 1:3
    @printf("%-8s %12.6f %12.6f %12.2e\n",
            buses[i], res.V_ang[i], ref_va[i],
            abs(res.V_ang[i] - ref_va[i]))
end

println("\nLine active power flows [MW]:")
line_names = ["L 0-1", "L 0-2", "L 1-2"]
# DC PF reference (from dc_power_flow.jl validation, baseMVA=100 convention)
dc_pf_flows = [266.67, 233.33, -33.33]  # same as AC for this lossless-like case
@printf("%-8s %10s %10s %10s %10s\n", "Line", "LACPF P", "Full AC P", "|ΔP|", "DC PF P")
for i in 1:3
    @printf("%-8s %10.4f %10.4f %10.4f %10.4f\n",
            line_names[i], res.P_flow[i], ref_pf[i],
            abs(res.P_flow[i] - ref_pf[i]), dc_pf_flows[i])
end

println("\nLine reactive power flows [MVAr]:")
@printf("%-8s %10s %10s %10s\n", "Line", "LACPF Q", "Full AC Q", "|ΔQ|")
for i in 1:3
    @printf("%-8s %10.4f %10.4f %10.4f\n",
            line_names[i], res.Q_flow[i], ref_qf[i],
            abs(res.Q_flow[i] - ref_qf[i]))
end

max_dvm = maximum(abs.(res.V_mag .- ref_vm))
max_dva = maximum(abs.(res.V_ang .- ref_va))
max_dp  = maximum(abs.(res.P_flow .- ref_pf))

println("\n" * "─"^50)
@printf("Max |ΔVm|:  %.2e p.u.\n", max_dvm)
@printf("Max |Δθ|:   %.2e rad\n",  max_dva)
@printf("Max |ΔP|:   %.4f MW\n",   max_dp)
println("─"^50)
println("Note: DC PF ignores R → no reactive power, Vm=1 always.")
println("LACPF recovers approximate Q flows and voltage deviations.")
println("="^60)

end # if abspath

File: python/linear_ac_pf.py
"""
Linearized AC Power Flow — Python reference implementation.
Validates against PyPSA full AC PF (network.pf()).
"""
import numpy as np
import pypsa
import warnings
import logging

warnings.filterwarnings("ignore")
logging.disable(logging.CRITICAL)


def solve_linear_ac_pf(buses, lines, generators, loads, baseMVA=100.0, v_nom=380.0):
    """
    Linearized AC Power Flow around flat start (|V|=1, theta=0).

    Solves:
        [ B' -G'] [Δθ ]   [P]
        [-G' -B'] [Δ|V|] = [Q]

    Parameters
    ----------
    buses      : list of bus names
    lines      : list of (from, to, r_pu, x_pu)
    generators : dict {bus: (P_MW, Q_MVAr)}
    loads      : dict {bus: (P_MW, Q_MVAr)}
    baseMVA    : system base [MVA]
    """
    n = len(buses)
    # r, x in physical Ω → convert to p.u.
    z_base = v_nom**2 / baseMVA    # 380²/100 = 1444 Ω
    G = np.zeros((n, n))
    B = np.zeros((n, n))

    for (f, t, r, x) in lines:
        f -= 1; t -= 1
        r_pu  = r / z_base
        x_pu  = x / z_base
        denom = r_pu**2 + x_pu**2
        g_ij  =  r_pu / denom
        b_ij  =  x_pu / denom      # positive |susceptance|

        G[f,f] += g_ij;  G[t,t] += g_ij
        B[f,f] += b_ij;  B[t,t] += b_ij
        G[f,t] -= g_ij;  G[t,f] -= g_ij
        B[f,t] -= b_ij;  B[t,f] -= b_ij

    G_mw = G * baseMVA
    B_mw = B * baseMVA

    P = np.zeros(n)
    Q = np.zeros(n)
    for bus, (p, q) in generators.items():
        P[bus-1] += p;  Q[bus-1] += q
    for bus, (p, q) in loads.items():
        P[bus-1] -= p;  Q[bus-1] -= q

    # Remove slack bus (index 0)
    idx = list(range(1, n))
    B_r = B_mw[np.ix_(idx, idx)]
    G_r = G_mw[np.ix_(idx, idx)]
    P_r = P[idx]
    Q_r = Q[idx]

    nm = n - 1
    A = np.block([[ B_r, -G_r],
                  [-G_r, -B_r]])
    rhs = np.concatenate([P_r, Q_r])
    x_sol = np.linalg.solve(A, rhs)

    dtheta = np.zeros(n);  dtheta[idx] = x_sol[:nm]
    dv     = np.zeros(n);  dv[idx]     = x_sol[nm:]
    V_mag  = 1.0 + dv
    theta  = dtheta

    P_flow, Q_flow = [], []
    for (f, t, r, x) in lines:
        f -= 1; t -= 1
        r_pu  = r / z_base
        x_pu  = x / z_base
        denom  = r_pu**2 + x_pu**2
        g_ij   = r_pu / denom
        b_ij   = x_pu / denom      # positive |susceptance|
        dth_ft = theta[f] - theta[t]
        dv_ft  = dv[f] - dv[t]
        P_flow.append(( b_ij * dth_ft + g_ij * dv_ft) * baseMVA)
        Q_flow.append(( b_ij * dv_ft  - g_ij * dth_ft) * baseMVA)

    return V_mag, theta, np.array(P_flow), np.array(Q_flow)


# ----------------------------------------------------------------
#  Full AC PF reference via PyPSA
# ----------------------------------------------------------------
def pypsa_ac_pf():
    net = pypsa.Network()
    net.set_snapshots([0])
    for i, name in enumerate(["Bus 0","Bus 1","Bus 2"]):
        net.add("Bus", name, v_nom=380.0)
    for k,(f,t,r,x) in enumerate([(1,2,0.01,0.1),(1,3,0.01,0.1),(2,3,0.01,0.1)]):
        net.add("Line", f"L{k}", bus0=f"Bus {f-1}", bus1=f"Bus {t-1}",
                r=r, x=x, s_nom=1e6)
    net.add("Generator","G0",bus="Bus 0",p_nom=500,control="Slack")
    net.add("Load","L1",bus="Bus 1",p_set=300)
    net.add("Load","L2",bus="Bus 2",p_set=200)
    net.pf()
    buses = ["Bus 0","Bus 1","Bus 2"]
    vm  = net.buses_t.v_mag_pu.loc[0, buses].values
    va  = net.buses_t.v_ang.loc[0, buses].values
    pf  = net.lines_t.p0.loc[0].values
    qf  = net.lines_t.q0.loc[0].values
    return vm, va, pf, qf


# ================================================================
SEP = "=" * 60
print(SEP)
print("LINEARIZED AC POWER FLOW  (Python reference)")
print(SEP)

buses = ["Bus 0","Bus 1","Bus 2"]
lines = [(1,2,0.01,0.1),(1,3,0.01,0.1),(2,3,0.01,0.1)]
generators = {1: (500.0, 0.0)}
loads      = {2: (300.0, 0.0), 3: (200.0, 0.0)}

vm_lac, va_lac, pf_lac, qf_lac = solve_linear_ac_pf(
    buses, lines, generators, loads)

print("\nRunning PyPSA full AC PF for reference...")
vm_ref, va_ref, pf_ref, qf_ref = pypsa_ac_pf()

print(f"\n{'Bus':<8} {'|V| LACPF':>10} {'|V| Full AC':>11} {'|Δ|':>10}")
print("-"*42)
for i in range(3):
    print(f"{buses[i]:<8} {vm_lac[i]:>10.6f} {vm_ref[i]:>11.6f} "
          f"{abs(vm_lac[i]-vm_ref[i]):>10.2e}")

print(f"\n{'Bus':<8} {'θ LACPF (rad)':>13} {'θ Full AC (rad)':>15} {'|Δ|':>10}")
print("-"*50)
for i in range(3):
    print(f"{buses[i]:<8} {va_lac[i]:>13.6f} {va_ref[i]:>15.6f} "
          f"{abs(va_lac[i]-va_ref[i]):>10.2e}")

lnames = ["L 0-1","L 0-2","L 1-2"]
print(f"\n{'Line':<8} {'P LACPF':>10} {'P Full AC':>10} {'|ΔP|':>8} "
      f"{'Q LACPF':>10} {'Q Full AC':>10} {'|ΔQ|':>8}")
print("-"*68)
for i in range(3):
    print(f"{lnames[i]:<8} {pf_lac[i]:>10.4f} {pf_ref[i]:>10.4f} "
          f"{abs(pf_lac[i]-pf_ref[i]):>8.4f} "
          f"{qf_lac[i]:>10.4f} {qf_ref[i]:>10.4f} "
          f"{abs(qf_lac[i]-qf_ref[i]):>8.4f}")

print(f"\nMax |ΔVm| = {max(abs(vm_lac-vm_ref)):.2e} p.u.")
print(f"Max |Δθ|  = {max(abs(va_lac-va_ref)):.2e} rad")
print(f"Max |ΔP|  = {max(abs(pf_lac-pf_ref)):.4f} MW")
print(f"Max |ΔQ|  = {max(abs(qf_lac-qf_ref)):.4f} MVAr")
print(SEP)
